app/post.go

```diff
@@ -402,13 +408,160 @@ func (a *App) SendEphemeralPost(userId string, post *model.Post) *model.Post {
 		post.Props = model.StringInterface{}
 	}
 
+	post.GenerateActionIds()
 	message := model.NewWebSocketEvent(model.WEBSOCKET_EVENT_EPHEMERAL_MESSAGE, "", post.ChannelId, userId, nil)
-	message.Add("post", a.PreparePostForClient(post, true).ToJson())
+	post = a.PreparePostForClient(post, true)
+	post = a.AddActionCookiesToPost(post)
+	message.Add("post", post.ToJson())
+	a.Publish(message)
+
+	return post
+}
+
+func (a *App) UpdateEphemeralPost(userId string, post *model.Post) *model.Post {
+	post.Type = model.POST_EPHEMERAL
+
+	post.UpdateAt = model.GetMillis()
+	if post.Props == nil {
+		post.Props = model.StringInterface{}
+	}
+
+	post.GenerateActionIds()
+	message := model.NewWebSocketEvent(model.WEBSOCKET_EVENT_POST_EDITED, "", post.ChannelId, userId, nil)
+
+	// TODO: why? but it doesn't work without clearing ChannelId, which is needed ^^.
+	post.ChannelId = ""
+
+	post = a.PreparePostForClient(post, true)
+	post = a.AddActionCookiesToPost(post)
+	message.Add("post", post.ToJson())
+	a.Publish(message)
+
+	return post
+}
+
+func (a *App) DeleteEphemeralPost(userId string, post *model.Post) *model.Post {
+	post.Type = model.POST_EPHEMERAL
+	post.UpdateAt = model.GetMillis()
+	message := model.NewWebSocketEvent(model.WEBSOCKET_EVENT_POST_DELETED, "", post.ChannelId, userId, nil)
+
+	message.Add("post", post.ToJson())
 	a.Publish(message)
 
 	return post
 }
 
+func (a *App) AddActionCookiesToPost(o *model.Post) *model.Post {
+	p := o.Clone()
+
+	// retainedProps carry over their value from the old post, including no value
+	retainPropKeys := []string{"override_username", "override_icon_url"}
+	retainProps := map[string]interface{}{}
```

Comment:
Can I get an explanation of what this "retain and remove props" stuff is?

Reply:
This was in the code before, I just added it to the cookie use-case. As I understand it, "override_username", "override_icon_url" values including nils are carried from the original post to the updated one, and so they need to go into the cookie. I could just use 2 fixed values/struct fields, but was not sure of the original intent there.